cibyl/orchestrator.py

```diff
@@ -102,9 +96,76 @@ def select_source_method(self, system, argument):
         return Source.get_source_method(system.name.value, system_sources,
                                         argument.func)
 
+    def _consistent_environment(self, env):
+        """Check if an environment should be used according to user input.
+
+        :param env: Model to validate
+        :type env: :class:`.Environment`
+        :returns: Whether the environment is consistent with user input
+        :rtype: bool
+        """
+        user_env = self.parser.ci_args.get("env_name")
+        if user_env:
+            return env.name.value in user_env.value
+        return True
+
+    def _consistent_system(self, system):
+        """Check if a system should be used according to user input.
+
+        :param system: Model to validate
+        :type system: :class:`.System`
+        :returns: Whether the system is consistent with user input
+        :rtype: bool
+        """
+        is_valid_system = True
+        name = system.name.value
+
+        user_system_type = self.parser.ci_args.get("system_type")
+        if user_system_type:
+            if system.system_type.value not in user_system_type.value:
+                is_valid_system = False
+
+        user_system_name = self.parser.ci_args.get("system_name")
+        if user_system_name:
+            if name not in user_system_name.value:
+                is_valid_system = False
+
+        user_systems = self.parser.ci_args.get("systems")
+        if user_systems:
+            if name not in user_systems.value:
+                is_valid_system = False
+        return is_valid_system
+
+    def consistent_with_user_input(self):
+        """Filter environments and systems according to user input.
+
+        :returns: Systems that can be used according to user input
+        :rtype: dict
+        """
+        user_systems = []
+        user_envs = []
+        for env in self.environments:
+            if not self._consistent_environment(env):
+                continue
+            user_envs.append(env)
+            for system in env.systems:
+                if self._consistent_system(system):
+                    user_systems.append(system)
+
+        if not user_envs:
+            raise NoValidEnvironment()
+        if not user_systems:
+            raise NoValidSystem()
+        # we keep only the environments consistent with the user input, this
+        # should be helpful for the publisher to avoid showing unnecessary
+        # information
+        self.environments = user_envs
+        return user_systems
+
     def run_query(self, start_level=1):
         """Execute query based on provided arguments."""
         last_level = -1
+        user_valid_systems = self.consistent_with_user_input()
```

Comment:
May be call it from `create_ci_environments`

Reply:
It can't be called from create_ci_environments because at that point we still have not parsed the user input, it only loads the configuration data